File: Twitter/Controller/Login/LoginViewController.m
//
//  LoginViewController.m
//  Twitter
//
//  Created by Дмитрий Матвеенко on 24/01/2019.
//  Copyright © 2019 GkFoxes. All rights reserved.
//

#import "LoginViewController.h"

@interface LoginViewController ()<UITextFieldDelegate>

@property (weak, nonatomic) IBOutlet UITextField *emailTextField;
@property (weak, nonatomic) IBOutlet UITextField *passwordTextField;

- (IBAction)loginTapeed:(id)sender;

@property (strong, nonatomic) FIRDatabaseReference *databaseRef;

@end

@implementation LoginViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.databaseRef = [[FIRDatabase database] reference];
}

- (void)viewDidAppear:(BOOL)animated {
    [_emailTextField becomeFirstResponder];
}

// Choose next TextField or Login, when click Return key
- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    
    if (textField == _emailTextField) {
        [textField resignFirstResponder];
        [_passwordTextField becomeFirstResponder];
    } else if (textField == _passwordTextField) {
        [self loginTapeed:self];
    }
    
    return YES;
}

// Warning alert message
- (void) showAlertMessageWithTitle:(NSString *)title message:(NSString *)message {
    UIAlertController * alert = [UIAlertController alertControllerWithTitle:title message:message preferredStyle:UIAlertControllerStyleAlert];
    UIAlertAction* defaultAction = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * action) {}];
    
    [alert addAction:defaultAction];
    [self presentViewController:alert animated:YES completion:nil];
}

// MARK: - Login Button Action
- (IBAction)loginTapeed:(id)sender {
    
    if (_emailTextField.text.length > 0 && _passwordTextField.text.length >= 6) {
        
        [[FIRAuth auth] signInWithEmail:self->_emailTextField.text password:self->_passwordTextField.text completion:^(FIRAuthDataResult * _Nullable authResult, NSError * _Nullable error) {
                                 
            if (error == nil) {
                [[[[self.databaseRef child:@"user_profiles"] child:authResult.user.uid] child:@"handle"] observeSingleEventOfType:FIRDataEventTypeValue withBlock:^(FIRDataSnapshot * _Nonnull snapshot) {
                    
                    if (snapshot.exists) {
                        
                        //Setting Side Menu
                        AppDelegate *appDelegate = (AppDelegate*)[[UIApplication sharedApplication]delegate];
                        [appDelegate settingFeedSideMenu];
                        
                        [self performSegueWithIdentifier:@"feedFromLoginSegue" sender:sender];
                    } else {
                        [self performSegueWithIdentifier:@"registerDetailFromLogin" sender:sender];
                    }
                }];
            } else {
                [self showAlertMessageWithTitle:@"Something Wrong" message:@"Please try again later."];
            }
        }];
    } else {
        [self showAlertMessageWithTitle:@"Empty field" message:@"You did not fill all the fields or your password is small, please check again."];
    }
}

@end

File: twitter-project/Controller/Login/LoginViewController.swift
//
//  LoginViewController.swift
//  twitter-project
//
//  Created by Дмитрий Матвеенко on 11.07.2018.
//  Copyright © 2018 Дмитрий Матвеенко. All rights reserved.
//

import UIKit
import Firebase

class LoginViewController: UIViewController {
    
    var reference: DatabaseReference!
    
    let shared = SharedManager.shared
    
    @IBOutlet weak var loginLabel: UILabel!
    @IBOutlet weak var emailTextField: UITextField!
    @IBOutlet weak var passwordTextField: UITextField!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        reference = Database.database().reference(withPath: "users")
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        emailTextField.text = ""
        passwordTextField.text = ""
    }
    
    override func viewDidAppear(_ animated: Bool){
        super.viewDidAppear(animated)
        
        if Auth.auth().currentUser != nil {
            shared.isLogin = true
        }
        
        emailTextField.becomeFirstResponder()
    }
    
    // MARK: - Warning alert manager
    
    func alertWarning(title: String, message: String) {
        let alert = UIAlertController(title: title, message: message, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
        self.present(alert, animated: true, completion: nil)
    }
    
    // MARK: - Button Action
    
    @IBAction func loginTapped(_ sender: UIButton) {
        guard let email = emailTextField.text, let password = passwordTextField.text,
            email != "", password != "" else {
                alertWarning(title: "Empty field", message: "You did not fill all the fields, please check again.")
                return
        }
        
        let queue = DispatchQueue.global(qos: .userInteractive)
        queue.async{
            Auth.auth().signIn(withEmail: email, password: password, completion: { (user, error) in
                if error != nil {
                    self.alertWarning(title: "Can not login", message: "Warning, please check the entered data.")
                }
                
                if user != nil {
                    self.shared.isLogin = true
                    self.shared.isLoginFirst = true
                    
                    DispatchQueue.main.async {
                        self.performSegue(withIdentifier: "feedSegue", sender: nil)
                    }
                } else {
                    self.alertWarning(title: "Can not login", message: "User does not exist, please check the entered data.")
                }
            })
        }
    }
}
